drivers/video/ov7670.c

```diff
@@ -0,0 +1,482 @@
+	{OV7670_GAM7, 0x80},
+	{OV7670_GAM8, 0x88},
+	{OV7670_GAM9, 0x8f},
+	{OV7670_GAM10, 0x96},
+	{OV7670_GAM11, 0xa3},
+	{OV7670_GAM12, 0xaf},
+	{OV7670_GAM13, 0xc4},
+	{OV7670_GAM14, 0xd7},
+	{OV7670_GAM15, 0xe8},
+	/* Histogram-based AEC/AGC Control */
+	{OV7670_HAECC1, 0x78},
+	{OV7670_HAECC2, 0x68},
+	{OV7670_HSYEN, 0xff},
+	{0xa1, 0x03},
+	{OV7670_HAECC3, 0xdf},
+	{OV7670_HAECC4, 0xdf},
+	{OV7670_HAECC5, 0xf0},
+	{OV7670_HAECC6, 0x90},
+	/* Automatic black Level Compensation */
+	{0xb0, 0x84},
+	{0xb1, 0x0c},
+	{0xb2, 0x0e},
+	{0xb3, 0x82},
+	{0xb8, 0x0a},
+};
+
+static int ov7670_get_caps(const struct device *dev, enum video_endpoint_id ep,
+			   struct video_caps *caps)
+{
+	caps->format_caps = fmts;
+	return 0;
+}
+
+static int ov7670_set_fmt(const struct device *dev, enum video_endpoint_id ep,
+			  struct video_format *fmt)
+{
+	const struct ov7670_config *config = dev->config;
+	struct ov7670_data *data = dev->data;
+	uint8_t com7 = 0U;
+	uint8_t i = 0U;
+
+	if (fmt->pixelformat != VIDEO_PIX_FMT_RGB565 && fmt->pixelformat != VIDEO_PIX_FMT_YUYV) {
+		LOG_ERR("Only RGB565 and YUV supported!");
+		return -ENOTSUP;
+	}
+
+	if (!memcmp(&data->fmt, fmt, sizeof(data->fmt))) {
+		/* nothing to do */
+		return 0;
+	}
+
+	memcpy(&data->fmt, fmt, sizeof(data->fmt));
+
+	if (fmt->pixelformat == VIDEO_PIX_FMT_RGB565) {
+		com7 |= 0x4;
+	}
+
+	/* Set output resolution */
+	while (fmts[i].pixelformat) {
+		if (fmts[i].width_min == fmt->width && fmts[i].height_min == fmt->height &&
+		    fmts[i].pixelformat == fmt->pixelformat) {
+			/* Set output format */
+			switch (fmts[i].width_min) {
+			case 176: /* QCIF */
+				com7 |= BIT(3);
+				break;
+			case 320: /* QVGA */
+				com7 |= BIT(4);
+				break;
+			case 352: /* CIF */
+				com7 |= BIT(5);
+				break;
+			default: /* VGA */
+				break;
+			}
+			/* Program COM7 to set format */
+			return i2c_reg_write_byte_dt(&config->bus, OV7670_COM7, com7);
+		}
+		i++;
+	}
+	LOG_ERR("Unsupported format");
+	return -ENOTSUP;
+}
+
+static int ov7670_get_fmt(const struct device *dev, enum video_endpoint_id ep,
+			  struct video_format *fmt)
+{
+	struct ov7670_data *data = dev->data;
+
+	if (fmt == NULL) {
+		return -EINVAL;
+	}
+	memcpy(fmt, &data->fmt, sizeof(data->fmt));
+	return 0;
+}
+
+static int ov7670_init(const struct device *dev)
+{
+	const struct ov7670_config *config = dev->config;
+	int ret, i;
+	uint8_t pid;
+	struct video_format fmt;
+	const struct ov7670_reg *reg;
+
+	if (!i2c_is_ready_dt(&config->bus)) {
+		/* I2C device is not ready, return */
+		return -ENODEV;
+	}
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios)
+	/* Power up camera module */
+	if (gpio_is_ready_dt(&config->pwdn)) {
+		ret = gpio_pin_configure_dt(&config->pwdn, GPIO_OUTPUT_INACTIVE);
+		if (ret < 0) {
+			LOG_ERR("Could not clear power down pin: %d", ret);
+			return ret;
+		}
+	}
+#endif
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
+	/* Reset camera module */
+	if (gpio_is_ready_dt(&config->reset)) {
```

Comment:
If the reset gpio is not ready, the sensor will not intialize, we shouldn't continue and should `return -ENODEV` ?

Reply:
Good catch- I've updated this check